### PR TITLE
Convert CollectionType, SpecialFolderType to enum

### PR DESCRIPTION
**Pull request by crobibero**
Requested by client devs

Draft pending testing

### CODE REVIEW

File: MediaBrowser.Controller/Entities/UserViewBuilder.cs [hunk 1]
```diff
@@ -952,7 +952,7 @@ private BaseItem[] GetMediaFolders(User user, IEnumerable<string> viewTypes)
                     {
                         var folder = i as ICollectionFolder;
 
-                        return folder is not null && viewTypes.Contains(folder.CollectionType ?? string.Empty, StringComparison.OrdinalIgnoreCase);
+                        return folder?.CollectionType != null && viewTypes.Contains(folder.CollectionType.Value);
```
Comment (Shadowghost) [inline MediaBrowser.Controller/Entities/UserViewBuilder.cs:955]:
```suggestion
                        return folder?.CollectionType is not null && viewTypes.Contains(folder.CollectionType.Value);
```

File: Emby.Server.Implementations/Library/Resolvers/SpecialFolderResolver.cs [hunk 2]
```diff
@@ -78,7 +79,8 @@ private string GetCollectionType(ItemResolveArgs args)
                     }
                 })
                 .Select(i => _fileSystem.GetFileNameWithoutExtension(i))
-                .FirstOrDefault();
+                .Select(i => Enum.TryParse<CollectionType>(i, out var collectionType) ? collectionType : (CollectionType?)null)
+                .FirstOrDefault(i => i != null);
```
Comment (Shadowghost) [inline Emby.Server.Implementations/Library/Resolvers/SpecialFolderResolver.cs:83]:
```suggestion
                .FirstOrDefault(i => i is not null);
```

File: MediaBrowser.Controller/Library/ItemResolveArgs.cs [hunk 3]
```diff
@@ -220,7 +221,7 @@ public bool ContainsFileSystemEntryByName(string name)
             return GetFileSystemEntryByName(name) != null;
```
Comment (Shadowghost) [inline MediaBrowser.Controller/Library/ItemResolveArgs.cs:221]:
```suggestion
            return GetFileSystemEntryByName(name) is not null;
```

File: Jellyfin.Data/Enums/CollectionType.cs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+    /// <summary>
+    /// Tv genres collection.
+    /// </summary>
+    TvGenres = 102,
+
+    /// <summary>
+    /// Tv genre collection.
+    /// </summary>
+    TvGenre = 103,
```
Comment (nielsvanvelzen) [inline Jellyfin.Data/Enums/CollectionType.cs:86]: What's the difference between these two?

File: Jellyfin.Data/Enums/CollectionType.cs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+    /// <summary>
+    /// Movie genres collection.
+    /// </summary>
+    MovieGenres = 114,
+
+    /// <summary>
+    /// Movie genre collection.
+    /// </summary>
+    MovieGenre = 115
```
Comment (nielsvanvelzen) [inline Jellyfin.Data/Enums/CollectionType.cs:146]: What's the difference between these two?

File: Jellyfin.Data/Enums/CollectionType.cs (hunk 4 above)
Comment (crobibero) [reply Jellyfin.Data/Enums/CollectionType.cs:86]: Yay legacy code where they added support for both…

Comment (Shadowghost) [reply Jellyfin.Data/Enums/CollectionType.cs:86]: Can't we write a migration and get rid of them?

Comment (EraYaN) [reply Jellyfin.Data/Enums/CollectionType.cs:86]: We could mark one as `[Obsolete]` or something right?

Comment (crobibero) [reply Jellyfin.Data/Enums/CollectionType.cs:86]: The difference is actually as follows so it needs to stay... I've opted to just hide them from the api spec

https://github.com/jellyfin/jellyfin/blob/master/MediaBrowser.Controller/Entities/UserViewBuilder.cs#L73:L77

Comment (EraYaN) [reply Jellyfin.Data/Enums/CollectionType.cs]: Aah so one is really a TvGenreList (list of TVGenres) and the other one a list of items in a specific genre.

**Comment by crobibero** [comment]
@jellyfin-bot rebase

**Review by Bond-009 — APPROVED** [review]

**Review by Shadowghost — APPROVED** [review]

File: Jellyfin.Data/Attributes/OpenApiIgnoreEnumAttribute.cs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+using System;
+
+namespace Jellyfin.Data.Attributes;
+
+/// <summary>
+/// Attribute to specify that the enum value is to be ignored when generating the openapi spec..
```
Comment (barronpm) [inline Jellyfin.Data/Attributes/OpenApiIgnoreEnumAttribute.cs:6]:
```suggestion
/// Attribute to specify that the enum value is to be ignored when generating the openapi spec.
```

**Review by barronpm — APPROVED** [review]

**Pull request merged**